source/server/listener_manager_impl.cc

```diff
@@ -165,41 +163,30 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, const std::st
         factory.createFilterFactoryFromProto(Envoy::ProtobufWkt::Empty(), *this));
   }
 
-  // Skip lookup and update of the SSL Context if there is only one filter chain
-  // and it doesn't enforce any SNI restrictions.
-  const bool skip_context_update =
-      (config.filter_chains().size() == 1 &&
-       config.filter_chains()[0].filter_chain_match().sni_domains().empty());
-
-  absl::optional<uint64_t> filters_hash;
-  uint32_t has_tls = 0;
-  uint32_t has_stk = 0;
+  bool install_tls_inspector = false;
+  std::set<std::size_t> filter_chains;
   for (const auto& filter_chain : config.filter_chains()) {
-    std::vector<std::string> sni_domains(filter_chain.filter_chain_match().sni_domains().begin(),
-                                         filter_chain.filter_chain_match().sni_domains().end());
-    if (!filters_hash) {
-      filters_hash = RepeatedPtrUtil::hash(filter_chain.filters());
-      filter_factories_ =
-          parent_.factory_.createNetworkFilterFactoryList(filter_chain.filters(), *this);
-    } else if (filters_hash.value() != RepeatedPtrUtil::hash(filter_chain.filters())) {
-      throw EnvoyException(fmt::format("error adding listener '{}': use of different filter chains "
-                                       "is currently not supported",
+    const auto& filter_chain_match = filter_chain.filter_chain_match();
+
+    std::size_t filter_chain_hash = MessageUtil::hash(filter_chain_match);
+    if (filter_chains.find(filter_chain_hash) != filter_chains.end()) {
```

Comment:
In the case of a matching hash, I think we should do a full equality-compare, in case of a hash-collision. Which at that point, just add the filter_chain_match itself to the set/hash-table and try to find that.

Reply:
Done (please check if that's what you've meant), but I find it very unlikely to have hash collision here.